_posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md

```diff
@@ -0,0 +1,104 @@
+---
+title: 'Stop robots and crawlers causing errors in your Rails application'
+description: 'Beware the remorseless robot army.'
+layout: article
+category: ruby
+image:
+  base: '2018/stop-robots-crawlers-triggering-errors-rails'
+  alt: 'BB-8'
+  source: 'https://unsplash.com/photos/C8VWyZhcIIU'
+  credit: 'Joseph Chan'
+---
+
+When a Rails application can't find a record, it throws a 404 error. This is a standard HTTP code for browsers meaning 'not found'.
+
+When you have an Internet-facing site various search engines will be crawling it. As you change things, certain URLs might change or cease to exist. This means search engines/crawlers can start generating a lot of ‘not found’ errors by trying to load pages that used to exist.
+
+
+## Instead of…
+
+...getting a bunch of unhelpful, distracting noise in your monitoring setup, or errors in your logs, when Google (or another web crawler) hits deleted public pages…
+
+
+### Or...
+
+...naively swallowing all your 404 errors.
+
+```ruby
+class ApplicationController < ActionController::Base
+  rescue_from ActiveRecord::RecordNotFound, with: :not_found
+
+  private
+
+  def not_found(exception)
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+
+## Use…
+
+...the [`is_crawler`](https://github.com/ccashwell/is_crawler) gem and the configure it like so:
+
+### `application_controller.rb`
+
+```ruby
+class ApplicationController < ActionController::Base
+  include IsCrawler
+
+  rescue_from ActiveRecord::RecordNotFound, with: :not_found
+
+  private
+
+  def not_found(exception)
+    if Rails.env.production? && is_crawler?(request.user_agent)
+      render_404
+    else
+      raise exception
+    end
+  end
+
+  def render_404
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+One of my sites gets hit by many more crawlers that aren't included by default in the gem. So I add these to the list of crawlers in an initializer:
```

Comment:
You either have to say something like: 'The 'is_crawler' gem comes with many default crawlers'..then the 'many more crawlers' phrasing works..

Or you change the sentence to something like: "One of my sites gets hit by many crawlers that aren't included.." Otherwise the 'more' comparison doesn't make sense.